FOCUS #5: Resource ID spec changes and supporting documents (#42)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: supporting_content/dimensions/resourceid.md

> # Dimension: ResourceId

Reviewer: ```suggestion
# Column: ResourceId
```

## Example mappings

> | Provider  | Data set                 | Vendor column                        |

Reviewer: Should this be either "Provider column" or just "Column"?
```suggestion
| Provider  | Data set                 | Column                        |
```

|-----------|--------------------------|--------------------------------------|
| AWS       | CUR                      | line_item_resource_id                |
| GCP       | Big Query Billing Export | resource.name? resource.global_name? |
| Microsoft | Cost details             | ResourceId                           |

> ## Examples

Reviewer: Do we need to qualify that these are existing examples and not suggested values when implementing FOCUS?

Author: Section name says Example Mappings which should be clear hopefully.

| AWS       | Some cases, just a resource-id is provided        | I-0b21f4c4434558933 OR my-cur-bucket                                                             |
| AWS       | Tax, API call charges etc.                        | (null)                                                                                           |
| GCP       | Resource based                                    | //container.googleapis.com/projects/<project_id>/locations/<location>/clusters/<cluster>         |
| GCP       | Non-resource based (e.g. network egress)          | (null)                                                                                           |
| Microsoft | Resource group resources                          | /subscriptions/<guid>/resourceGroups/<name>/providers/<provider>/<resource-type>/<resource-name> |
| Microsoft | Subscription resources                            | /subscriptions/<guid>/providers/<provider>/<resource-type>/<resource-name>                       |
| Microsoft | Tenant resources                                  | /providers/<provider>/<resource-type>/<resource-name>                                            |

> | Microsoft | Purchase charges                                  | (null)                                                                                           |

Reviewer: nit: We might want to move this down after reservation and savings plan purchases, since it's the catch-all bucket.
```suggestion
| Microsoft | Marketplace and other purchases                                  | (null)                                                                                           |
```

| Microsoft | Reservation purchases                             | /providers/Microsoft.Capacity/reservationOrders/<guid>                                           |

> | Microsoft | Savings plan purchases                            | /providers/Microsoft.Capacity/savingsPlanOrders/<guid>                                           |

Reviewer: ```suggestion
| Microsoft | Savings plan purchases                            | /providers/Microsoft.BillingBenefits/savingsPlanOrders/<guid>                                           |
```

### Documentation

AWS:

> > <https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html>

Reviewer: nit: Should this be in Markdown format?

> Interesting note on the "The resource identifier" which is at the end of the ARN: This is the name of the resource, the ID of the resource, or a [resource path](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html#arns-paths). Some resource identifiers include a parent resource (sub-resource-type/parent-resource/sub-resource) or a qualifier such as a version (resource-type:resource-name:qualifier)."

Reviewer: Do we care to define ARN?

Author: I don't think so. This is really meant to capture scratch notes IMO. It's not going in the spec. I was considering not even converting to markdown to allow for it to be updated easily without needing to convert back-and-forth.

I've updated to indicate this is from the link above (which is official doc on ARN so should be sufficient)

> Microsoft:

Reviewer: https://learn.microsoft.com/azure/cost-management-billing/automate/understand-usage-details-fields

GCP:
> <https://cloud.google.com/apis/design/resource_names> <https://cloud.google.com/asset-inventory/docs/resource-name-format> 

> ## Discussion / Scratch space

Reviewer: nit: Should we move action items to issues and potential discussion points that don't have a specific action to discussions? Not sure if we want to commit these to the repo, where they won't be tracked/closed.

Author: My goal was to provide a sneak peek for the things we discussed / considered as a part of building this. It's not really to track anything. Agree on those needing to be issues in the repo.

Author: Cleaned up more of the discussion items

Reviewer: Hmm... if it's minutes, we're missing a bunch. I thought this was the open items we didn't close, which we should be tracking somewhere... 🤔

Author: It's not minutes. This is just things we discussed and/or considered while building this. Minutes aren't a part of the PR. These are so people can get a sense for what was discussed and considered as a part of building the spec content.

If anything requires decisions beyond this work or needs to be tracked, they should be opened as an issue in the repo with the discussion topic label.

- Use one column for resource ID and make strong recommendations for the value to be globally unique within the provider. Eg. Full ARN for AWS.

> - Should this be prefixed with 'Vendor' e.g. VendorResourceID to show that it's a 1:1 mapping to a vendor provided column. 

Reviewer: I believe this should be "provider", not "vendor".
```suggestion
- Should this be prefixed with 'Provider' e.g. ProviderResourceID to show that it's a 1:1 mapping to a provider-defined column. 
```

- Should governance say "Nullable where cost is not associated with a resource"

- Do we make any statements about uniqueness? Or is this where we request a uniqueness composite key to be specified in a schema document

> - Can uniqueness and can it change over time?

Reviewer: Is this "Can we guarantee uniqueness..."? I don't remember.

> - Empty values should not be allowed if a user/systematically generated value is null (should this be an attribute specified as a higher level attribute OR within this scope)?

Reviewer: Can we remove this given the null handling rules define empty values are not allowed?

Author: Same as above, i think the goal is to show the things we discussed (sausage making process). Happy to discuss and change if we feel this doc / discussions space should be more buttoned down.
